### ISSUE TITLE
Remove the default value for '--data-template', stop auto-deleting `--data-directory`

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#1191]
Shadow's `--data-template` option copies a template directory to the data directory before running an experiment. By default, this is set to `shadow.data.template`. If someone's not familiar with this feature of Shadow, this default behaviour might not be very intuitive. Instead we should not set a default data template directory so that the user must set it explicitly.

For backwards-compatibility reasons, if the `--data-template` option was not set but a `shadow.data.template` directory exists, we should log a warning to let the user know that this default behaviour changed.

-------

We discussed removing this option entirely and requiring the user to copy the template directory manually before running the experiment, and also to not automatically delete the `shadow.data` directory before running the experiment, since this could be considered outside of Shadow's responsibility. But there are some other problems to consider. For example, someone could run an experiment, and then immediately run it again, but since the files from the previous experiment would still exist, the results from the second experiment might be different which could be confusing to the user. It's not immediately obvious which approach is best.

**Comment by robgjansen** [comment]
@sporksmith @stevenengler How about this:
- remove the `--data-template` option entirely
- remove the code in shadow that deletes the `shadow.data` directory (if it exists) before running the sim

I think the above should remove the unintuitive "magical" behavior. Then, to deal with the issue of accidentally running an experiment with a stale data directory:
- when starting shadow, if the `shadow.data` directory already exists, log a critical message and refuse to start
- add a new `--force` option, that will tell shadow, no really, I set up the `shadow.data` directory and I want you to run over the top of it

This seem more intuitive to me. Shadow never takes a destructive or magical action unless you tell it to.

**Comment by stevenengler** [comment]
> @sporksmith @stevenengler How about this:
>
> [...]

This sounds okay to me. One downside could be that it was almost impossible before before to run Shadow over top of an existing experiment, but now it becomes possible to with this `--force` option. This would be a user error, but it's not an issue that was possible before. (Are there reasons for ever running Shadow over top of an existing experiment? For example maybe to bootstrap a tor network in order to make future experiments faster (consensus and server descriptors will already be downloaded)?)

A third option could be:

- *keep* the `--data-template` option
- remove the code in shadow that deletes the `shadow.data` directory (if it exists) before running the sim
- when starting shadow, if the `--data-directory` (i.e., `shadow.data`) already exists, log a critical message and refuse to start

The advantage is that you could never accidentally run over top of an existing Shadow experiment, while still never taking a destructive action, but the downside is that we would have to keep the `--data-template` option.

Anyways I'm okay with any of these options.

**Comment by robgjansen** [comment]
> Are there reasons for ever running Shadow over top of an existing experiment?

My thinking is that the `--force` option would take the place of the template option - it would be used for cases like Tor where we have some pre-initialized host home directories that we want to start the simulation with (containing the fingerprints and keys).

So, outside of shadow I would set up the data directories for the hosts to contain fingerprints, keys, torrc files, etc. Then I start shadow with `--force`. In other words, my workflow for large sims would be:

```
cp -ar shadow.data.template shadow.data
shadow --force ... shadow.config.xml
```

But for other experiments where there is no pre-initialization data, you should just run:

```
shadow --force ... shadow.config.xml
```

The more I think about this the more I am in favor of NOT having shadow delete or copy directories. `rm` and `cp` do this much more efficiently anyway.

**Comment by stevenengler** [comment]
Okay sounds good. If @sporksmith also agrees, do you want to copy your comment to a new issue and we can close this one? As for implementing any of these config/cli features, I want to wait until we have the Shadow tests running in the CI before we start making these changes.

**Comment by sporksmith** [comment]
As @stevenengler pointed out, if you *do* have a template, it's easier in this model to accidentally run in a dirty directory, since you have to use `--force`, so shadow won't stop you. IIUC the correct workflow for (repeatedly) running a tor sim would actually be more like:

```
rm -rf shadow.data
cp -ar shadow.data.template shadow.data
shadow --force ... shadow.config.xml
```

I moderately prefer the plan of keeping `--data-directory`, since it makes it easier to use a template and ensure you don't accidentally run in a dirty directory. I do like the idea of dropping the default value and forcing the user to explicitly specify it.

In either model I'm ok with no longer removing `shadow.data` and failing if it's already there. I haven't been bitten this but I could easily imagine being extremely frustrated by accidentally wiping out the results of an hours-long simulation. I might call `--force` something a bit more specific like `--allow-dirty-data-dir`. If we do change this behavior, we'll need to change our tests to delete the data dir before running, but that's easy enough. (Or maybe we don't even need an override option if we keep the `--data-directory` option?)

**Comment by robgjansen** [comment]
> I moderately prefer the plan of keeping --data-directory

I think you mean `--data-template`.

OK, let's keep the `--data-template` option, and then I agree that we don't need a `--force`/`--allow-dirty-data-dir` option.

Then I think we've converged on the following behavior:

- Keep the `--data-directory` option; it defaults to `shadow.data`.
- Shadow will not delete the `--data-directory` in any case. If the `--data-directory` already exists, Shadow will log a warning and refuse to start.
- Keep the `--data-template` option, but make it off by default.
- If `--data-template` is given and the dir exists, copy it to the `--data-directory` path and then start the sim.
- If `--data-template` is given and the dir does not exist, log a warning and refuse to start.

In this case, I think we just need these changes:
- [ ] Change `--data-template` to be off by default
- [ ] If `--data-directory` exists, log warning and refuse to start instead of deleting the directory contents.

**Comment by robgjansen** [comment]
> do you want to copy your comment to a new issue and we can close this one?

Let's just keep this issue open. I'll update the name of the issue.

**Comment by stevenengler** [comment]
For the changes, I think we also want:

- [ ] If the `--data-template` option was not set but a `shadow.data.template` directory exists, log a warning and continue.
- [ ] If `--data-template` is given and the dir does not exist, log a warning and refuse to start.

**Comment by robgjansen** [comment]
> If the --data-template option was not set but a shadow.data.template directory exists, log a warning and continue.

Actually, this one is not needed: if `--data-template` is not set, then we have no idea which directory we should be checking to print a warning. (`--data-template` is the option that tells us the name of the template dir; we don't hard code `shadow.data.template`).

**Comment by sporksmith** [comment]
It seems like potentially a useful warning, since otherwise users' old scripts that invoke shadow with an implicit data dir will now silently not use it. OTOH there are enough other breaking changes in phantom that maybe it's sufficient to just note it there.

Speaking of which - maybe we should start a draft of the phantom release notes, including breaking changes?

**Comment by robgjansen** [comment]
> there are enough other breaking changes in phantom

This. We are not done messing with the config yet either. I could see many of the other option names changing yet as we work on https://github.com/shadow/shadow/milestone/27, which I strongly believe we should finish before releasing 2.0.

**Comment by stevenengler** [comment]
I think we covered everything we wanted to in this issue by #1210.
